Migrate to ESBuild in Webpack

ShGKme opened the pull request:
### ☑️ Resolves

#### Part 1: fixing issues with `webrtc-adapter` and `attachmediastream`.

`webrtc-adapter` library is supposed to be imported as ESModule and doesn't work with `require`.

Because of this, we had a special processing for `webrtc-adapter` with `babel-plugin-add-module-export`:

https://github.com/nextcloud/spreed/blob/36f53c1a44d2176f666fa81a0d2736c41687daac/webpack.common.config.js#L46-L65

But this problem can be avoided by:
1. Importing `webrtc-adapter` with ESM import instead of `require`
2. Using `attachmediastream` with its build for bundlers (
   ```js
   import attachmediastream from 'attachmediastream/attachmediastream.bundle.js'`
   ```

This PR does as described above. So we don't need additional babel plugin and special processing for `webrtc-adapter`. This is also required for migration to ESBuild as well as Vite.

**This change can be separated into individual PR.**

cc @danxuliu 

#### Part 2: Use ESBuild for transpiling JS instead of Terser and minimizing instead of Terser

* Fix: #9406

Production build with Terser minification in Webpack is very slow and requires much RAM.
The same (but not so huge) problem with Babel.

I close all applications on my 16GB RAM server to be able to make a production build 😣

ESBuild is slightly worse as a minifier but much more resource-effective. Results on my server:

Metric | Babel + Terser | ESBuild
---|---|---
`npm run build` assets size | 51.33 MB | 52.41 MB
`npm run build` time | 83.9s | 26.5s
`npm run build` RAM | 8.8 GB | 3.3 GB
`npm run dev` time | 25.5s | 13.8s
`npm run dev` RAM | 1.3 GB | 0.9 GB

### 👀 Alternative solutions

Fully migrate to Vite

### 🚧 Tasks

- [x] Add `esbuild-loader`
- [x] Use `ESBuildPlugin` as minimizer
- [x] Use `esbuild-loader` as JavaScript transpiler
- [x] Setup config merging rules
- [x] Check with Talk Desktop
  - Only requires to add `esbuild-loader` as a dependency and change target to es2022
  - But requires a bit more work to make a config compatible with both esbuild and babel
- [x] Update Babel config to be used only for Jest tests

### 🏁 Checklist

- [x] ⛑️ Tests (unit and/or integration) are included or not required
- [x] 📘 API documentation in `docs/` has been updated or is not required
- [x] 📗 User documentation in https://github.com/nextcloud/documentation/tree/master/user_manual/talk has been updated or is not required
- [x] 🔖 Capability is added or not needed

ShGKme commented:
We can use this as a simple solution to save some time and RAM requirements for production build.

Or we can try to make a full Vite migration and make a hackable shared Vite config with @susnux

ShGKme commented:
UPD: this vite config https://github.com/nextcloud/nextcloud-vite-config

ShGKme commented:
[Vite migration](https://github.com/nextcloud/spreed/pull/10065) is not doing well for me for now.

I'd consider using this PR as a very simple and not so risky solution.

SystemKeeper commented:
Works for me only if I downgrade `esbuild-loader` to version 2:

```diff
diff --git a/package.json b/package.json
index 8520ca615..79ce7f82d 100644
--- a/package.json
+++ b/package.json
@@ -78,7 +78,7 @@
     "@vue/vue2-jest": "^29.2.4",
     "babel-loader-exclude-node-modules-except": "^1.2.1",
     "babel-plugin-add-module-exports": "^1.0.4",
-    "esbuild-loader": "^3.0.1",
+    "esbuild-loader": "^2.21.0",
     "flush-promises": "^1.0.2",
     "jest": "^29.6.1",
     "jest-environment-jsdom": "^29.6.1",
diff --git a/webpack.config.js b/webpack.config.js
index 450b878a2..b87670a7c 100644
--- a/webpack.config.js
+++ b/webpack.config.js
@@ -1,6 +1,6 @@
 const path = require('node:path')
 
-const { EsbuildPlugin } = require('esbuild-loader')
+const { ESBuildMinifyPlugin } = require('esbuild-loader')
 const webpack = require('webpack')
 const { mergeWithRules } = require('webpack-merge')
 
@@ -62,7 +62,7 @@ module.exports = talkWebMerge(nextcloudWebpackConfig, commonWebpackConfig, {
 
        optimization: {
                minimizer: [
-                       new EsbuildPlugin({
+                       new ESBuildMinifyPlugin({
                                target: 'es2020',
                        }),
                ],
```

If I use version 3, I get

```
Error: File '@vue/tsconfig/tsconfig.json' not found.
    at Oe (/Users/marcel/Desktop/Nextcloud/workspace/server/apps-extra/spreed/node_modules/get-tsconfig/dist/index.cjs:3:9061)
    at G (/Users/marcel/Desktop/Nextcloud/workspace/server/apps-extra/spreed/node_modules/get-tsconfig/dist/index.cjs:3:9848)
    at Object.je [as getTsconfig] (/Users/marcel/Desktop/Nextcloud/workspace/server/apps-extra/spreed/node_modules/get-tsconfig/dist/index.cjs:3:10586)
    at Object.ESBuildLoader (/Users/marcel/Desktop/Nextcloud/workspace/server/apps-extra/spreed/node_modules/esbuild-loader/dist/index.cjs:49:23)
    at LOADER_EXECUTION (/Users/marcel/Desktop/Nextcloud/workspace/server/apps-extra/spreed/node_modules/loader-runner/lib/LoaderRunner.js:132:14)
    at runSyncOrAsync (/Users/marcel/Desktop/Nextcloud/workspace/server/apps-extra/spreed/node_modules/loader-runner/lib/LoaderRunner.js:133:4)
    at iterateNormalLoaders (/Users/marcel/Desktop/Nextcloud/workspace/server/apps-extra/spreed/node_modules/loader-runner/lib/LoaderRunner.js:250:2)
    at /Users/marcel/Desktop/Nextcloud/workspace/server/apps-extra/spreed/node_modules/loader-runner/lib/LoaderRunner.js:223:4
    at /Users/marcel/Desktop/Nextcloud/workspace/server/apps-extra/spreed/node_modules/webpack/lib/NormalModule.js:836:15
    at eval (eval at create (/Users/marcel/Desktop/Nextcloud/workspace/server/apps-extra/spreed/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:12:1)
```

It works if I add a empty `tsconfig.json` file...

SystemKeeper commented:
Okay, so the issue is the following:

* I have cloned the spreed repo in `server/apps-extra/spreed/`. 
* According to https://www.typescriptlang.org/docs/handbook/tsconfig-json.html the `tsconfig.json` file is searched in the current directory and if not found, searched in the parent directory
* Since we don't have a `tsconfig.json` in talk, the `tsconfig.json` of server is used
* As I usually don't develop frontend on server, I did not run `npm ci` in the server repository. As the `tsconfig.json` from server extends `@vue/tsconfig/tsconfig.json`, this file can't be found on my machine.
* Running `npm ci` inside the server repo was enough to make talk in a subdirectory build (note: it worked fine as well when moved outside of the server repository, e.g. no `tsconfig.json` could be located in any parent directory).

Would it make sense to adjust the behaviour in talk about that? Either webpack config or include a `tsconfig.json` to prevent any sideeffects with parent-level `tsconfig.json` files?

SystemKeeper commented:
For the performance:

```
# npm run build
webpack 5.76.2 compiled with 2 warnings in 15411 ms

# npm run dev
webpack 5.76.2 compiled successfully in 7790 ms
```

so that's quite nice ;)

ShGKme commented:
Thank you @SystemKeeper for your investigation!

So it seems that `eslint-loader` is trying to find `tsconfig` by `get-tsconfig` and then the config is found in the server root. And there is no way to disable it ><

https://github.com/esbuild-kit/esbuild-loader/blob/60cb3b4746c675167ff3a3ab4f4bddefde9e44fe/src/loader.ts#L51-L65

I'll implicitly set some config to avoid such errors.

ShGKme commented:
> Reference on master:
> # npm run build
> 49627 ms
> # npm run dev
> 12890 ms

So fast 😯

ShGKme commented:
@SystemKeeper Could you, please, check that now it works for you?

SystemKeeper commented:
> @SystemKeeper Could you, please, check that now it works for you?

Works fine now for me, thanks 👍

ShGKme commented:
Ready for the the review, description updated

Antreesy reviewed (approved):
Worked for me as well, finally can build a Talk app :smiling_face_with_tear:

nickvergessen reviewed (approved):
Also compiles here, 11s dev build, 21s prod build

ShGKme commented:
Calls doesn't work without any errors in the console. 🥲

danxuliu reviewed (changes requested):
> Calls doesn't work without any errors in the console. :smiling_face_with_tear:

Firefox shows `Uncaught TypeError: Peer is not a constructor` and the trace points to [_src/utils/simplewebrtc/webrtc.js_](https://github.com/nextcloud/spreed/blob/0adc9730877942e6d3762ea79178aeff284c1c78/src/utils/webrtc/simplewebrtc/webrtc.js#L102), which loads the constructor with [`const Peer = require('./peer.js')`](https://github.com/nextcloud/spreed/blob/0adc9730877942e6d3762ea79178aeff284c1c78/src/utils/webrtc/simplewebrtc/webrtc.js#L9). Maybe the change from `require` to `import` needs to be propagated as well to the rest of the files?

ShGKme commented:
> Firefox shows `Uncaught TypeError: Peer is not a constructor` and the trace points to [_src/utils/simplewebrtc/webrtc.js_](https://github.com/nextcloud/spreed/blob/0adc9730877942e6d3762ea79178aeff284c1c78/src/utils/webrtc/simplewebrtc/webrtc.js#L102), which loads the constructor with [`const Peer = require('./peer.js')`](https://github.com/nextcloud/spreed/blob/0adc9730877942e6d3762ea79178aeff284c1c78/src/utils/webrtc/simplewebrtc/webrtc.js#L9). Maybe the change from `require` to `import` needs to be propagated as well to the rest of the files?

Thanks, that helped! Indeed the problem was around ESM-CJS interop.

Possible solutions are:
1. Use only one module system (ESM) everywhere and do not mix
2. Implicitly import `default` as it is done [here](https://github.com/nextcloud/spreed/blob/4a7aedd0391bf27f0fc3e170d07016a3997bff3c/src/utils/webrtc/simplewebrtc/localmedia.js#L10-L16)

I'd go with the first one to avoid further interop issues, for consistency and possible migration to Vite (that works on top of ESM)

susnux commented:
> Use only one module system (ESM) everywhere and do not mix

Much cleaner solution :+1:

ShGKme commented:
Calls work now, including video, virtual backgrounds, and blur.

Antreesy commented on src/utils/webrtc/simplewebrtc/getscreenmedia.js:
```diff
@@ -23,7 +23,13 @@ const getUserMedia = function(constraints, callback) {
-module.exports = function(mode, constraints, cb) {
+/**
+ *
+ * @param mode
```
Will be super to add some docs

Code that might be related:
https://github.com/HenrikJoreteg/webrtc.js/blob/f7da40fcc599e5650c210f7864b4ebeeefbbc31b/webrtc.bundle.js

Antreesy reviewed (approved):
Also tested the main functionality in calls and chats.
We could try to merge it and see if we'll catch some bugs (because of these changes) during the development or backport (though any issues with bundling will be seen on sermo next day, I suppose)

nickvergessen commented:
> or backport (though any issues with bundling will be seen on sermo next day, I suppose).

This is far from a change I would put into a stable branch …

Antreesy commented:
> This is far from a change I would put into a stable branch …

I meant, that it could bring changes we will notice when backporting another fixes. I also don't mean to backport exactly this PR, and let it simmer on the master branch

danxuliu commented on src/utils/webrtc/simplewebrtc/localmedia.js:
```diff
@@ -1,26 +1,24 @@
+import TrackEnabler from '../../media/pipeline/TrackEnabler.js'
+import TrackToStream from '../../media/pipeline/TrackToStream.js'
+import VirtualBackground from '../../media/pipeline/VirtualBackground.js'
+import * as webrtcIndex from '../index.js'
```
When `require('../index.js')` was used all the exported elements had to be accessed through the `webrtcIndex` object. However, only `mediaDevicesManager` is used in the whole file, so now it would be possible to import just that variable and change the use of `webrtcIndex.mediaDevicesManager` to `mediaDevicesManager`.

danxuliu reviewed (commented):
Tested and works :+1:

As a rebase is anyway needed I would suggest to squash the last commit, `build: always use ESM and don't mix with CJS`, into the first one (and adjust the commit description if needed), `build: fix bundling issues with webrtc-adapter and attachmediastream `, as calls do not work after the first commit without the changes from the last commit.

Besides that, when `import attachMediaStream from 'attachmediastream` was used, [a mock for `attachmediastream` was needed in the unit tests](https://github.com/nextcloud/spreed/blob/5cb20b992b131a9427c5e125d4fa3005bff708a8/src/__mocks__/attachmediastream.js#L2-L3). Now that the bundle is used the mock is no longer needed (I have not actually checked why, though; maybe because the bundle includes its own copy of `webrtc-adapter` or something like that, but I do not really know), so it could be removed.

ShGKme commented:
- Rebased onto `master`, including conflict with https://github.com/nextcloud/spreed/pull/10230
- Removed the mock for `attachmediastream` mentioned by @danxuliu 
- Squashed fixups

danxuliu reviewed (commented):
> > As a rebase is anyway needed I would suggest to squash the last commit, `build: always use ESM and don't mix with CJS`, into the first one (and adjust the commit description if needed), `build: fix bundling issues with webrtc-adapter and attachmediastream `, as calls do not work after the first commit without the changes from the last commit.
> 
> Not exactly. They are different issues.
> 
> The first commit fixes the issue we had even in Webpack, having to add the Babel plugin as a fix.
> 
> While moving from mixed modules to ESM is required for ESBuild (and Vite in the future), because it has different module interop mechanism.

I have to admit that I forgot that changing `require('XXX')` calls to `require('XXX').default` calls [was another option to fix the first commit](https://github.com/nextcloud/spreed/pull/10055#issuecomment-1665514421) (and, in fact, changing only `require('./peer.js')` would be needed, not the others). That is why I thought that moving from mixed modules to ESM was also a requirement to fix the issues in Webpack :facepalm:

Had I remembered it before I would have suggested to change `require('./peer.js')` to `require('./peer.js').default` in the first commit instead ;-) (but it would still be a nice change for correctness and future reference :-) )

In any case, tested and works :+1:

danxuliu reviewed (approved):
As agreed with @ShGKme , rebased to fix the first and last commits.

Tested and works :+1:
